File: firmware/src/scoreboard/nba.py
"""
NBA game data model and binary wire-format deserialization.

Firmware mirror of the backend's `NbaGame` domain model
(`backend/src/nba/types.rs`) and its wire encoding — the NORMATIVE spec is
the `scoreboard_wire::nba` module of `crates/scoreboard-wire`.
Model classes are plain-attribute value types,
treated as immutable after construction, read by the display thread at
20 FPS — same contract as scoreboard/mlb.py.

Unlike soccer, the game clock travels as ESPN's display string ("10:08",
"53.0" under a minute): a basketball stop-clock cannot be extrapolated
between polls (there is no clock-running signal), so the display re-renders
the string each poll and `phase` says when it is meaningless (breaks).

Shared value types and wire primitives come from scoreboard.wire — same
contract as scoreboard/mlb.py and scoreboard/soccer.py.
"""

import struct

from .wire import (
    DeserializeError,
    GAME_STATE_IN,
    GAME_STATE_POST,
    GAME_STATE_PRE,
    HDR_SIZE,
    LastPlay,
    PregameTeam,
    TeamColors,
    TeamState,
    check_version,
    dispatch_detail,
    read_str,
)

# Live fixed section (offset 2): flags, period, phase, away/home score u16,
# then away/home color pairs u32.
_LIVE_FMT = "<BBBHHIIII"
_LIVE_SIZE = struct.calcsize(_LIVE_FMT)  # 23

# Pregame fixed section (offset 2): flags, away wins/losses u16, home
# wins/losses u16, start_time u32, then color pairs.
_PRE_FMT = "<BHHHHIIIII"
_PRE_SIZE = struct.calcsize(_PRE_FMT)  # 29

# Final fixed section (offset 2): periods_played, away linescore len, home
# linescore len, away/home score u16, then color pairs (same layout as the
# MLB final, quarters in place of innings).
_FINAL_FMT = "<BBBHHIIII"
_FINAL_SIZE = struct.calcsize(_FINAL_FMT)  # 23

_FLAG_LAST_PLAY = 0x01

_PRE_FLAG_AWAY_RECORD = 0x01
_PRE_FLAG_HOME_RECORD = 0x02

# Live phase codes (wire byte at offset 4): breaks render without a clock.
PHASE_IN_PROGRESS = 0
PHASE_HALFTIME = 1
PHASE_END_OF_PERIOD = 2


def period_name(period: int) -> str:
    """Display name of a period: Q1-Q4, then OT / 2OT / ..."""
    if period <= 4:
        return "Q" + str(period)
    if period == 5:
        return "OT"
    return str(period - 4) + "OT"


class LiveGame:
    """Top-level live NBA snapshot (`/basketball/nba/games/{id}`).

    `clock` is ESPN's display string, exact at fetch time and never
    extrapolated; `phase` is a PHASE_* code (the clock reads "0.0" or a
    reset "12:00" during breaks, so the phase is the only render signal).
    """

    wire_state = GAME_STATE_IN

    def __init__(
        self,
        game_id: str,
        period: int,
        clock: str,
        phase: int,
        home: TeamState,
        away: TeamState,
        last_play: LastPlay | None,
    ) -> None:
        self.game_id = game_id
        self.period = period
        self.clock = clock
        self.phase = phase
        self.home = home
        self.away = away
        self.last_play = last_play

    @classmethod
    def from_struct(cls, buf) -> "LiveGame":
        """Parse an NBA live payload (see crates/scoreboard-wire)."""
        end = len(buf)
        check_version(buf, end)
        if end < HDR_SIZE + _LIVE_SIZE:
            raise DeserializeError(
                "@2", f"truncated fixed section: {end} < {HDR_SIZE + _LIVE_SIZE}"
            )

        (
            flags, period, phase,
            away_score, home_score,
            away_primary, away_alternate, home_primary, home_alternate,
        ) = struct.unpack_from(_LIVE_FMT, buf, HDR_SIZE)

        if phase > PHASE_END_OF_PERIOD:
            raise DeserializeError("@4", f"invalid live phase code: {phase}")

        o = HDR_SIZE + _LIVE_SIZE  # 25
        game_id, o = read_str(buf, o, end, "game_id")
        away_abbr, o = read_str(buf, o, end, "away abbreviation")
        home_abbr, o = read_str(buf, o, end, "home abbreviation")
        clock, o = read_str(buf, o, end, "clock")

        last_play: LastPlay | None = None
        if flags & _FLAG_LAST_PLAY:
            play_id, o = read_str(buf, o, end, "last play id")
            play_text, o = read_str(buf, o, end, "last play text")
            last_play = LastPlay(play_id, play_text)

        if o != end:
            raise DeserializeError(f"@{o}", f"{end - o} unexpected trailing bytes")

        return cls(
            game_id=game_id,
            period=period,
            clock=clock,
            phase=phase,
            home=TeamState(home_abbr, home_score, TeamColors(home_primary, home_alternate)),
            away=TeamState(away_abbr, away_score, TeamColors(away_primary, away_alternate)),
            last_play=last_play,
        )


class PregameGame:
    """Upcoming NBA game.

    Duck-typed to the contract `state.set_pregame` reads (venue /
    weather_temp / weather_condition / start_epoch / away / home with
    wins-losses-pitcher), so the NBA pregame screen reuses the whole MLB
    pregame pipeline: real venue and records, and the fields basketball
    never has (weather, probables) are permanently absent — the renderer
    already omits absent fields.
    """

    wire_state = GAME_STATE_PRE

    def __init__(self, game_id: str, start_epoch: int, venue: str,
                 home: PregameTeam, away: PregameTeam) -> None:
        self.game_id = game_id
        self.start_epoch = start_epoch
        self.venue = venue
        self.weather_temp = None
        self.weather_condition = None
        self.home = home
        self.away = away

    @classmethod
    def from_struct(cls, buf) -> "PregameGame":
        """Parse an NBA pregame payload (see crates/scoreboard-wire)."""
        end = len(buf)
        check_version(buf, end)
        if end < HDR_SIZE + _PRE_SIZE:
            raise DeserializeError(
                "@2", f"truncated fixed section: {end} < {HDR_SIZE + _PRE_SIZE}"
            )

        (
            flags,
            away_wins, away_losses, home_wins, home_losses,
            start_time,
            away_primary, away_alternate, home_primary, home_alternate,
        ) = struct.unpack_from(_PRE_FMT, buf, HDR_SIZE)

        o = HDR_SIZE + _PRE_SIZE  # 31
        game_id, o = read_str(buf, o, end, "game_id")
        away_abbr, o = read_str(buf, o, end, "away abbreviation")
        home_abbr, o = read_str(buf, o, end, "home abbreviation")
        venue, o = read_str(buf, o, end, "venue")

        if o != end:
            raise DeserializeError(f"@{o}", f"{end - o} unexpected trailing bytes")

        away_has_record = bool(flags & _PRE_FLAG_AWAY_RECORD)
        home_has_record = bool(flags & _PRE_FLAG_HOME_RECORD)

        return cls(
            game_id=game_id,
            start_epoch=start_time,
            venue=venue,
            away=PregameTeam(
                away_abbr,
                TeamColors(away_primary, away_alternate),
                away_wins if away_has_record else None,
                away_losses if away_has_record else None,
                None,
            ),
            home=PregameTeam(
                home_abbr,
                TeamColors(home_primary, home_alternate),
                home_wins if home_has_record else None,
                home_losses if home_has_record else None,
                None,
            ),
        )


class FinalTeam:
    """One team's final snapshot: abbreviation, colors, total, line score.

    `line` is per-quarter points (quarter 1 first, overtime periods after),
    copied out of the response buffer with `bytes(...)` so it survives the
    next request.
    """

    def __init__(self, abbreviation: str, colors: TeamColors, score: int,
                 line: bytes) -> None:
        self.abbreviation = abbreviation
        self.colors = colors
        self.score = score
        self.line = line


class FinalGame:
    """Top-level final NBA snapshot."""

    wire_state = GAME_STATE_POST

    def __init__(self, game_id: str, periods_played: int, home: FinalTeam,
                 away: FinalTeam) -> None:
        self.game_id = game_id
        self.periods_played = periods_played
        self.home = home
        self.away = away

    @classmethod
    def from_struct(cls, buf) -> "FinalGame":
        """Parse an NBA final payload (see crates/scoreboard-wire)."""
        end = len(buf)
        check_version(buf, end)
        if end < HDR_SIZE + _FINAL_SIZE:
            raise DeserializeError(
                "@2", f"truncated fixed section: {end} < {HDR_SIZE + _FINAL_SIZE}"
            )

        (
            periods_played, away_len, home_len,
            away_score, home_score,
            away_primary, away_alternate, home_primary, home_alternate,
        ) = struct.unpack_from(_FINAL_FMT, buf, HDR_SIZE)

        o = HDR_SIZE + _FINAL_SIZE  # 25
        if o + away_len + home_len > end:
            raise DeserializeError(
                f"@{o}",
                f"truncated linescores: need {away_len + home_len} bytes, "
                f"have {end - o}",
            )
        # Copy out of the shared response buffer: the source memoryview
        # aliases the client's reusable _response_buf and is clobbered by the
        # next request.
        away_line = bytes(buf[o:o + away_len])
        o += away_len
        home_line = bytes(buf[o:o + home_len])
        o += home_len

        game_id, o = read_str(buf, o, end, "game_id")
        away_abbr, o = read_str(buf, o, end, "away abbreviation")
        home_abbr, o = read_str(buf, o, end, "home abbreviation")

        if o != end:
            raise DeserializeError(f"@{o}", f"{end - o} unexpected trailing bytes")

        return cls(
            game_id=game_id,
            periods_played=periods_played,
            home=FinalTeam(
                home_abbr, TeamColors(home_primary, home_alternate), home_score, home_line
            ),
            away=FinalTeam(
                away_abbr, TeamColors(away_primary, away_alternate), away_score, away_line
            ),
        )


def parse_game_detail(buf) -> "LiveGame | PregameGame | FinalGame":
    """Parse an NBA detail payload, dispatching on the state header byte."""
    return dispatch_detail(buf, LiveGame, PregameGame, FinalGame)

File: crates/scoreboard-wire/src/nba.rs
//! NBA game detail (`GET /basketball/nba/games/{game_id}`).
//!
//! # Live (state = 1)
//!
//! Fixed 23-byte section at offset 2 (`<BBBHHIIII`):
//!
//! | off | type | field                                            |
//! |-----|------|--------------------------------------------------|
//! | 2   | u8   | flags (bit0 = last play present)                 |
//! | 3   | u8   | period (quarter 1–4; overtime = 5+)              |
//! | 4   | u8   | phase (0=in progress, 1=halftime, 2=end of period) |
//! | 5   | u16  | away_score                                       |
//! | 7   | u16  | home_score                                       |
//! | 9   | u32  | away_colors.primary (0x00RRGGBB)                 |
//! | 13  | u32  | away_colors.alternate                            |
//! | 17  | u32  | home_colors.primary                              |
//! | 21  | u32  | home_colors.alternate                            |
//!
//! Strings from offset 25: `game_id`, `away.abbreviation`,
//! `home.abbreviation`, `clock` (display-shaped: "10:08"; "53.0" under a
//! minute; meaningless during breaks — render by `phase`; never extrapolated,
//! the firmware re-renders it each poll), then **iff** bit0: `last_play.id`
//! (change-detection key for the flash) and `last_play.text`.
//!
//! # Pregame (state = 0)
//!
//! Fixed 29-byte section at offset 2 (`<BHHHHIIIII`):
//!
//! | off | type | field                                            |
//! |-----|------|--------------------------------------------------|
//! | 2   | u8   | flags (bit0 = away record, bit1 = home record)   |
//! | 3   | u16  | away_wins  (0 if no record)                      |
//! | 5   | u16  | away_losses                                      |
//! | 7   | u16  | home_wins                                        |
//! | 9   | u16  | home_losses                                      |
//! | 11  | u32  | start_time (unix epoch, seconds, UTC)            |
//! | 15  | u32  | away_colors.primary                              |
//! | 19  | u32  | away_colors.alternate                            |
//! | 23  | u32  | home_colors.primary                              |
//! | 27  | u32  | home_colors.alternate                            |
//!
//! Strings from offset 31, in order: `game_id`, `away.abbreviation`,
//! `home.abbreviation`, `venue`.
//!
//! # Final (state = 2)
//!
//! The shared final layout (see [`crate::common`]) with per-quarter points in
//! the line scores; overtime extends both past 4.

use crate::codec::{Sink, SinkExt, color_pairs, le_u16, le_u32};
use crate::common::{
    FinalParts, payload_reader, read_final, read_header, write_final, write_header,
};
use crate::error::{BufferFull, DecodeError, DecodeErrorKind};
use crate::{FinalTeam, GameState, HEADER_LEN, LastPlay, LivePhase, Record, TeamColors, TeamState};

const LIVE_LEN: usize = 23;
const PRE_LEN: usize = 29;

const FLAG_LAST_PLAY: u8 = 0x01;

const PRE_FLAG_AWAY_RECORD: u8 = 0x01;
const PRE_FLAG_HOME_RECORD: u8 = 0x02;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Game<'a> {
    Pregame(Pregame<'a>),
    Live(Live<'a>),
    Final(Final<'a>),
}

impl Game<'_> {
    pub fn state(&self) -> GameState {
        match self {
            Game::Pregame(_) => GameState::Pregame,
            Game::Live(_) => GameState::Live,
            Game::Final(_) => GameState::Final,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Live<'a> {
    pub game_id: &'a str,
    /// Quarter 1–4; overtime periods pass through as 5+.
    pub period: u8,
    pub phase: LivePhase,
    /// Display-shaped and exact at poll time — never extrapolated.
    pub clock: &'a str,
    pub away: TeamState<'a>,
    pub home: TeamState<'a>,
    /// Absent before the opening tip.
    pub last_play: Option<LastPlay<'a>>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Pregame<'a> {
    pub game_id: &'a str,
    /// Unix epoch seconds, UTC.
    pub start_time: u32,
    pub venue: &'a str,
    pub away: PregameTeam<'a>,
    pub home: PregameTeam<'a>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct PregameTeam<'a> {
    pub abbreviation: &'a str,
    pub colors: TeamColors,
    pub record: Option<Record>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Final<'a> {
    pub game_id: &'a str,
    /// 4, or more with overtime.
    pub periods_played: u8,
    pub away: FinalTeam<'a>,
    pub home: FinalTeam<'a>,
}

pub fn encode<S: Sink + ?Sized>(game: &Game<'_>, out: &mut S) -> Result<(), BufferFull> {
    write_header(out, game.state())?;
    match game {
        Game::Pregame(pregame) => write_pregame(out, pregame),
        Game::Live(live) => write_live(out, live),
        Game::Final(game) => write_final(
            out,
            game.game_id,
            game.periods_played,
            &game.away,
            &game.home,
        ),
    }
}

pub fn decode(buf: &[u8]) -> Result<Game<'_>, DecodeError> {
    match read_header(buf)? {
        GameState::Pregame => read_pregame(buf).map(Game::Pregame),
        GameState::Live => read_live(buf).map(Game::Live),
        GameState::Final => read_final(buf).map(|parts| {
            let FinalParts {
                game_id,
                periods,
                away,
                home,
            } = parts;
            Game::Final(Final {
                game_id,
                periods_played: periods,
                away,
                home,
            })
        }),
    }
}

fn write_live<S: Sink + ?Sized>(out: &mut S, game: &Live<'_>) -> Result<(), BufferFull> {
    out.u8(if game.last_play.is_some() {
        FLAG_LAST_PLAY
    } else {
        0
    })?;
    out.u8(game.period)?;
    out.u8(game.phase.code())?;
    out.u16(game.away.score)?;
    out.u16(game.home.score)?;
    out.color_pairs(game.away.colors, game.home.colors)?;

    out.string(game.game_id)?;
    out.string(game.away.abbreviation)?;
    out.string(game.home.abbreviation)?;
    out.string(game.clock)?;
    if let Some(play) = &game.last_play {
        out.string(play.id)?;
        out.string(play.text)?;
    }
    Ok(())
}

fn read_live(buf: &[u8]) -> Result<Live<'_>, DecodeError> {
    let mut reader = payload_reader(buf);
    let fixed = reader.fixed::<LIVE_LEN>()?;
    let flags = fixed[0];
    let phase = LivePhase::from_code(fixed[2]).ok_or_else(|| {
        DecodeError::at(
            HEADER_LEN + 2,
            DecodeErrorKind::InvalidCode {
                field: "live phase",
                code: fixed[2],
            },
        )
    })?;
    let (away_colors, home_colors) = color_pairs(fixed, 7);

    let game_id = reader.string("game_id")?;
    let away_abbreviation = reader.string("away abbreviation")?;
    let home_abbreviation = reader.string("home abbreviation")?;
    let clock = reader.string("clock")?;
    let last_play = if flags & FLAG_LAST_PLAY != 0 {
        Some(LastPlay {
            id: reader.string("last play id")?,
            text: reader.string("last play text")?,
        })
    } else {
        None
    };
    reader.finish()?;

    Ok(Live {
        game_id,
        period: fixed[1],
        phase,
        clock,
        away: TeamState {
            abbreviation: away_abbreviation,
            score: le_u16(fixed, 3),
            colors: away_colors,
        },
        home: TeamState {
            abbreviation: home_abbreviation,
            score: le_u16(fixed, 5),
            colors: home_colors,
        },
        last_play,
    })
}

fn write_pregame<S: Sink + ?Sized>(out: &mut S, game: &Pregame<'_>) -> Result<(), BufferFull> {
    let mut flags = 0u8;
    if game.away.record.is_some() {
        flags |= PRE_FLAG_AWAY_RECORD;
    }
    if game.home.record.is_some() {
        flags |= PRE_FLAG_HOME_RECORD;
    }
    let away_record = game.away.record.unwrap_or(Record { wins: 0, losses: 0 });
    let home_record = game.home.record.unwrap_or(Record { wins: 0, losses: 0 });

    out.u8(flags)?;
    out.u16(away_record.wins)?;
    out.u16(away_record.losses)?;
    out.u16(home_record.wins)?;
    out.u16(home_record.losses)?;
    out.u32(game.start_time)?;
    out.color_pairs(game.away.colors, game.home.colors)?;

    out.string(game.game_id)?;
    out.string(game.away.abbreviation)?;
    out.string(game.home.abbreviation)?;
    out.string(game.venue)
}

fn read_pregame(buf: &[u8]) -> Result<Pregame<'_>, DecodeError> {
    let mut reader = payload_reader(buf);
    let fixed = reader.fixed::<PRE_LEN>()?;
    let flags = fixed[0];
    let (away_colors, home_colors) = color_pairs(fixed, 13);

    let game_id = reader.string("game_id")?;
    let away_abbreviation = reader.string("away abbreviation")?;
    let home_abbreviation = reader.string("home abbreviation")?;
    let venue = reader.string("venue")?;
    reader.finish()?;

    Ok(Pregame {
        game_id,
        start_time: le_u32(fixed, 9),
        venue,
        away: PregameTeam {
            abbreviation: away_abbreviation,
            colors: away_colors,
            record: (flags & PRE_FLAG_AWAY_RECORD != 0).then(|| Record {
                wins: le_u16(fixed, 1),
                losses: le_u16(fixed, 3),
            }),
        },
        home: PregameTeam {
            abbreviation: home_abbreviation,
            colors: home_colors,
            record: (flags & PRE_FLAG_HOME_RECORD != 0).then(|| Record {
                wins: le_u16(fixed, 5),
                losses: le_u16(fixed, 7),
            }),
        },
    })
}
